### ISSUE TITLE
Allow admin to hide the title on comments

### ISSUE THREAD

**Issue by stpaultim** [backdrop/backdrop-issues#4569]
**Description of the need**

The default situation in Backdrop CMS for comments is to display a title on all comments. If the user does not have the ability to manually create a custom title for the comment (this is an optional setting), then Backdrop CMS simply takes the first part of the first line and generates a title for the comment. 

This latter functionality doesn't seem very helpful to me. I believe that it should be possible to turn off the title on a comment field. 

**Proposed solutions**

We are currently discussing two different approaches towards resolving this issue and have two different PRs. 

4569-1  --  CSS Solution [edit: [PR 4061](https://github.com/backdrop/backdrop/pull/4061)]

This solution primarily relies on CSS and works pretty well with most core AND contrib themes (but not perfectly). 

- Update hook to replaces old comment_subject_field boolean setting with comment_title_options setting (3 choices - auto, custom, hidden)
- If "Hide Titles" is selected, a css class hide-comment-titles is added to the comment titles, it uses CSS to hide comment titles on any theme that is displaying comment titles and replaces them with a SVG "link" icon that acts as a permalink. 
- To accomplish css change in all contrib themes this PR adds a new css file at: core/modules/comment/css/comment_titles.css

4569-2  --  Template Solution [edit: [PR 4068](https://github.com/backdrop/backdrop/pull/4068)]

This solution updates the comment.tpl.php files for core, Bartik, and Basis and works with any contrib themes that inherit those templates/css. BUT, sites that use contrib templates that override comment.tpl.php files would not be able to hide comment titles unless their contrib theme is updated.

- Update hook to replaces old comment_subject_field boolean setting with comment_title_options setting (3 choices - auto, custom, hidden)
- If hidden is selected, it uses the comment.tpl.php file to hide the comment title in core, basis, and bartik.
- It does not have any effect on contrib themes using custom comment.tpl.php or comment.css files. Users will be unable to hide titles on those themes, unless the theme updates to support this feature.
- In BASIS, it uses CSS to add a permalink icon when the titles are hidden (see below).
- In core and Bartik, it keeps the same old permalink link that was there by default.

**Issues to Consider**

- By default comment titles also serve as a permalink. Simply hiding comment titles might leave some sites without any permalinks available for comments. Some core and contrib themes provide separate permalinks and this would not be a problem. 
- Because many contrib themes override the core comment.tpl.php and comment.css files, the first option above creates a NEW css file to make sure this solution works with as many contrib themes as possible.

**Additional information**

See [my comment](https://github.com/backdrop/backdrop-issues/issues/4413#issuecomment-678895796) in #4413 

> I am working on the website for Backdrop CMS Live event and I am running into this problem and it seems REALLY weird.
> 
> -1) "Allow comment title" is turned off (I see the point, that this might just be mislabeled).
> -2) My comments are all being auto populated by a title, which is really just the first line of text from the comment, formatted as a link which really does nothing but take me to the comment I'm already on.
> 
> This title and/or link looks really weird and doesn't serve any purpose that I can imagine. Also, I can't find anyway to hide this in the UI.
> 
> My screen shot comes from a default installation of Backdrop CMS with no changes to config, except enabling comments on this specific node and adding some comments.

![Your_first_post____Comments](https://user-images.githubusercontent.com/3144571/91006654-932d4600-e59f-11ea-94e8-51a3c078d468.png)

<hr />

To Do

- [ ] Decide if we should also add a new variable for permalink path. 

<hr />

**Next Steps**

The next step for this issue is to decide on the best approach (see options discussed above). 

Once we decide on an approach, we can work on the details of either PR and address issues with tests.

<hr />

- OLD PR by @VasasA https://github.com/backdrop/backdrop/pull/3280
- ADVOCATE: @stpaultim

**Comment by klonos** [comment]
In Drupal-land, https://www.drupal.org/project/comment_hide_subject does that, but it reuses this checkbox to determine if the title is to be hidden or not.

**Comment by VasasA** [comment]
I created a PR. It contains the "Hide comment title" option. Problem: Add-on themes may need to be updated.
(Note: The #4413 replaces the label "Allow comment title" with "Allow a custom title to be set for each comment".
https://github.com/backdrop/backdrop/commit/853b5e0b64cdd6a3ec2983b2fd9e975f7995808e )
![kép](https://user-images.githubusercontent.com/8567234/91775533-9651b300-ebeb-11ea-8889-dd07272f8f28.png)

**Comment by stpaultim** [comment]
Just a reminder that there are some changes to the "[] Allow comment title" option that were committed today.  https://github.com/backdrop/backdrop-issues/issues/4413 I can imagine that we might want to change that text once this is complete. 

Maybe "allow custom title" is not necessary and should be hidden if "hide comment title" is selected?

I think this option would have to be off by default - because that is current behavior. If it's off by default, it shouldn't effect existing sites very much.

**Comment by VasasA** [comment]
@stpaultim  I will create a new PR including solution of issue [4413](https://github.com/backdrop/backdrop-issues/issues/4413).
> Maybe "allow custom title" is not necessary and should be hidden if "hide comment title" is selected?

I think the "Allow a custom title to be set for each comment" option can also be useful when the "Hide comment title" is selected: You can collect the titles of comments without displaying and if you will make a new decision in the future, you will able to display all of the titles. I may be able to display a help text for title field if the "Hide comment title" is selected: "The title will not be displayed because the comment titles are hidden by the content editors." Is this help text useful?
![kép](https://user-images.githubusercontent.com/8567234/91812359-3e3d9f80-ec31-11ea-8b4a-f266f7de121c.png)

**Comment by stpaultim** [comment]
> I will create a new PR including solution of issue 4413.

Maybe you can use `git rebase` to pull in the changes to your branch? (or just a merge of the current dev branch).

> I think the "Allow a custom title to be set for each comment" option can also be useful when the "Hide comment title" is selected

Interesting idea. I can see how this might be useful, but I am a little worried that it might also be confusing. I would like to hear what some others think about this.

**Comment by VasasA** [comment]
I created a new [PR](https://github.com/backdrop/backdrop/pull/3284), because I had a problem with my repo. (The old PR is closed.)
It contains the commit of issue 4413 too, and a help text under "Hide comment title" field (1. screenshot). And another help text under the title field if the "Hide comment title" is selected. (2. screenshot)

![kép](https://user-images.githubusercontent.com/8567234/92036386-6849ab80-ed70-11ea-9053-57a3635841bb.png)
------------------------
![kép](https://user-images.githubusercontent.com/8567234/92036591-c4143480-ed70-11ea-8519-49359a40b4dd.png)

**Comment by bugfolder** [comment]
@stpaultim brought this issue to my attention, thx. There is a Drupal module, [Comment Hide Subject](https://www.drupal.org/project/comment_hide_subject) that does this. I have just started [porting it to Backdrop](https://github.com/backdrop-ops/contrib/issues/440).

**Comment by bugfolder** [comment]
Comment Hide Subject is now in Backdrop modules:

https://backdropcms.org/project/comment_hide_subject

**Comment by stpaultim** [comment]
I took some previous work by @VasasA and created a new PR. I'm still thinking about if this is the right approach or the implications of the issue @VasasA mentioned with contrib themes.

**Comment by stpaultim** [comment]
In the earlier PR by @VasasA, he was using some logic in the comment.tpl.php file which caused problems, because the solution would not work for any contrib themes using custom versions of this tpl.php file set up before this change was implemented. 

I looked at the solution in the "Comment Hide Subject" contrib module, which simply removes the value of the title during processing of comments. This seems to work, but I have no idea if there are other side-effects or reasons why this is a bad idea. 

I would love some feedback. 

``` 
// Check to see if title should be hidden
  if ($node_type->settings['comment_subject_hide']) {
    $variables['title'] = '';
  }
```
As far as I can tell, this is a simple code change that fixes a problem has vexes me every time I use comments.

**Comment by BWPanda** [comment]
It seems like in this vertical tab there are (at least) two types of settings:
1. Those that affect comment *forms* (adding/editing comments)
2. Those that affect comment *displays* (viewing comments under a node)

Perhaps, to avoid confusion, we should make two expanded fieldsets and move the various settings into them appropriately...

**Comment by argiepiano** [comment]
One problem of hiding the title of the comments is that you lose the permanent link to the comment. This link contains an anchor that allows you to reference the comment by pasting a URL. 

As a potential remedy: the Backdrop forum hides comment titles, but creates a "permalink" link that can be used to reference the comment. But this feature may be the topic for another issue.

> Perhaps, to avoid confusion, we should make two expanded fieldsets and move the various settings into them appropriately...

I personally don't find the configuration settings confusing.

I have tested this and it works for me. The code looks good to me too. But I think it'd be helpful to get others' opinions, especially given that this issue has been dormant for a long time and most people seem to be away today.

**Comment by docwilmot** [comment]
> One problem of hiding the title of the comments is that you lose the permanent link to the comment. 

I'd say thats an essential feature really, so I'm bumping to needs work. Perhaps we could do as GitHub issues do and use the date as the permalink.

**Comment by argiepiano** [comment]
Actually the core `comment.tpl.php` does provide an output for $permalink, but Basis removed that. The variable is still available to the template, though. So, a compromise for Basis may be to check if the title is empty. If so, maybe output the permalink variable. 

But then the issue is other contrib themes that override `comment.tpl.php`

**Comment by stpaultim** [comment]
I will try to work on this more later, but I welcome an alternate PR if anyone that is better at this than me is interested in this issue.

**Comment by stpaultim** [comment]
I have a simple idea (probably too simple). What if we replace the title with the permalink when the setting to "hide title" is enabled. 

![image](https://user-images.githubusercontent.com/3144571/166166380-101f0d94-06d3-4fc4-ac89-b8b24bfafca1.png)

To me, this looks much better than a machine generated title that often times matches the first line of the comment. In Bartik, some themes might already have a permalink link in the comment template, but since this is an optional setting, it is only an issue if the setting is activated, so it should be backward compatible.

**Comment by argiepiano** [comment]
That's not a bad idea. But how do people feel about having the permalink as <h3>? Unfortunately the <h3> is hardcoded in the template

**Comment by docwilmot** [comment]
If we permalink the date, then themes can decide if they want to substitute any text when the title is missing. That would be a quick fix.

**Comment by stpaultim** [comment]
@docwilmot - Thanks. Good note and easier than I thought it would be. At least this draft. Still not sure I've got the best solution, but it's working.

**Comment by stpaultim** [comment]
I had brief thoughts about getting this into 1.22, but I suspect that is not realistic - despite how close it is (or might be). 

@BWPanda What do you think of this?

![image](https://user-images.githubusercontent.com/3144571/166177720-3221b9fa-25da-491e-987e-94d942fcfc7f.png)

@docwilmot - This is what a comment looks like now after I change the date into a permalink instead of the title. 

![image](https://user-images.githubusercontent.com/3144571/166177904-b29124a6-319e-47b5-9a75-9470ee5c8e8e.png)

I believe that this is all backward compatible, because nothing happens unless you check the box, which is unchecked by default. If you check the box, two things happen:

1) The text is removed from the Title variable. 
2) The date created is converted to a permalink

This might not work well with some custom theme templates for comments, but I guess that is ok as long as this is optional. Bartik for example already has a dedicated permalink, so if you turn off titles you are left with two permalinks (the date and regular one). 

***Hide Titles***
![image](https://user-images.githubusercontent.com/3144571/166178114-88594293-1fef-4259-b7b9-f6bd4843f932.png)

But, if the box is not checked is also has two permalinks, but one of them is the title. 

***Default***
![image](https://user-images.githubusercontent.com/3144571/166178187-7fb7be56-8274-4c7b-b3a8-090da2ee1abb.png)

My biggest concern is the empty `<h3>` tags that are left in the markup when you check the box. I am not sure if that is a problem and/or if there is a better alternative.

**Comment by docwilmot** [comment]
> My biggest concern is the empty <h3> tags that are left in the markup when you check the box.

In `comment.tpl.php` you could wrap `$title` in a condition:
```
  <?php if ($title): ?>
    <h3><?php print $title; ?></h3>
  <?php endif; ?>
```
That should be safe, and makes sense since with or without this PR, you shouldnt want the `h3` tags showing up if no title.

**Comment by docwilmot** [comment]
Apart from the comment above and one other recommendation in the PR queue, this looks good. If any core committers are still awake, this shouldnt be too far from getting in.

**Comment by argiepiano** [comment]
> In `comment.tpl.php` you could wrap `$title` in a condition:

You'd need to do that both in the `core/modules/comment/templates` and in the template in the Basis theme.. and possibly Bartik. 

And then authors of contrib themes may also need to adjust their comment templates to adapt to the possibility of an empty title and some empty tags.

**Comment by stpaultim** [comment]
@docwilmot and @argiepiano - Thanks. The original PR by @VasasA did this. But, I looked around and realized that quite a few contrib themes seem to override the comment.tpl.php file which seems to be a problem. I've been trying to write this so that it does not depend upon login in the comment.tpl.php file. 

I suppose I could add that logic to the core template, Basis and Bartik - but also leave the current empty "h3". In that case, the ONLY time we would get an empty "h3" is if the theme overrides the comment.tpl.php file and even then the feature would work, it would just leave the empty "h3"'s.

Does this seem like better than what I have now?

**Comment by argiepiano** [comment]
Thinking a bit more about this. I think this is the path that would do the least "harm" to contrib themes:
- Do add the logic to core and Basis/Bartik comment template as above
- Change the docblock in those templates to indicate that $title can be empty when admin chooses that option in the content type configuration

Additionally, I'm not crazy about force-feeding a permalinked date. That can really mess up the look and design of contrib themes (more than an empty tag, I think).

So, how about this:
- Instead of making $created into a link, add a new variable in template_preprocess_comment called something like `created_permalinked` 
- Add logic to the template to print that variable instead of created if the title is empty
- Document the above

I think documentation is essential, as it will really give contrib themers a clue of what's going on

**Comment by stpaultim** [comment]
@argiepiano - Thanks for these ideas. I'm processing them. 

In the meantime, I did this at suggestion of @docwilmot. The custom title field is now hidden if the "hide titles" field gets checked. 

![image](https://user-images.githubusercontent.com/3144571/166182345-0b5af829-4c0d-4891-9efb-1f733b625dfa.png)

![image](https://user-images.githubusercontent.com/3144571/166182416-fc4a26a0-d054-491e-b893-493235b9e378.png)

**Comment by argiepiano** [comment]
But hiding the "Allow a custom title..." when it was checked can have unintended consequences. When it's hidden you really don't know if that hidden option is checked or unchecked.

**Comment by stpaultim** [comment]
> But hiding the "Allow a custom title..." can have unintended consequences. When it's hidden you really don't know if that hidden option is checked or unchecked.

I hear you. On the other hand, if its checked and the "allow custom titles" is hidden, it doesn't matter much anyway because titles are hidden. 

I suppose the real problem would be if someone "allow custom titles" is checked before "Hide titles". Then we have a situation in which users are being asked for custom titles that aren't being used and the admin might not be able to find the setting. 

OK, I'm convinced. I'll remove that again, unless someone speaks up. If by chance this PR does get approved. I think we could discuss that later.

**Comment by argiepiano** [comment]
Perhaps disabling the second checkbox? At least you can see if it's checked...

**Comment by stpaultim** [comment]
@argiepiano - I implemented your ideas. 
@docwilmot - I took out the conditional field based upon some legit feedback from @argiepiano. Maybe we discuss it later.

I think this is ready for additional testing and review.

**Comment by stpaultim** [comment]
The main outstanding question in my mind is whether to (1) automatically make the date created into a permalink or (2) to rely on the comment.tpl.php to provide a new permalink based upon creation date recognizing that many themes override this file. I checked the top ten themes (using the second option). 

Keep in mind - no site is effected unless they check this new option which is off by default.

Testing this with top 10 contrib themes (providing a new created_permalinked variable) - only 3 of these 10 themes are effected, IF the option is checked:

1) Bootstrap_Lite - Hides comment titles by default. Provides no permalink to comments. Unaffected by PR.
2) Corporate_Kiss - Provides it's own permalink - unaffected by PR
3) Lateral - Has it's own comment template. When comment titles are hidden, the permalink is lost.
4) Shasetsu - Provides no custom templates, should be unaffected by PR.
5) Pelerine - Has it's own comment template. When comment titles are hidden, the permalink is lost.
6) Tatsu - Does not override comment template, should be unaffected by PR. 
7) Snazzy - Has it's own comment template. When comment titles are hidden, the permalink is lost.
8) Materialize - Provides it's own permalink - unaffected by PR
9) Colihaut - Provides it's own permalink - unaffected by PR
10) Borg - Hides comment titles by default. Provides it's own permalink. Unaffected by PR.

**Comment by indigoxela** [comment]
Two concerns:

1. Only the word "Permalink" may not be optimal re A11Y - it's always the same text in the whole comment thread, and it gives no information about *what* is going to be permalinked.
2. Can we please find a solution that does not force me yet another time to rewrite all my contrib themes? (and keep in mind that there can also be custom themes or subthemes fiddling with comments)

**Comment by stpaultim** [comment]
>  it gives no information about what is going to be permalinked

I've updated the notes I made about the 10 contrib themes to hopefully be a bit more clear about what the permalink is in each case. 

For the purposes of this discussion there are three possibilities that are being used as permalinks. 

1) The default in core and most themes is that the comment title also serves as a permalink. 
2) Some themes create a permalink in the comments.tpl.php file with the text "permalink."
3) Our best available option, by default is to make the "date posted" into a permalink. This is standard way of permalinking posts on most social media platforms. 

> Can we please find a solution that does not force me yet another time to rewrite all my contrib themes?

We're doing our best? Do you have any suggestions?

This change will not force you to do anything with your themes. If you choose to make it possible for folks to turn off comment titles and still have permalinks, that's great. But, it's hardly a requirement. Our most popular contrib theme does not provide any permalinks with their comments. 

> and keep in mind that there can also be custom themes or subthemes fiddling with comments

Anyone that has created or modified a custom theme or sub-theme has already decided whether they want comments titles available or not and made whatever accommodations were necessary. 

In core, I think we try extremely hard to make sure we don't do anything that will break the work that folks have already done in their custom themes. But, I don't think it's possible for us to anticipate and avoid breaking things they MIGHT do in the future.

**Comment by indigoxela** [comment]
> ... are you simply opposed to fixing this problem in core at all?

@stpaultim sorry, but where did you read that in my concerns?

I'm aware that comment display (and handling) has quite a potential for improvement, but changes to defaults (templates) have to get handled with care - so they're a real improvement.

What worried me here was that "last minute" is was attempted to rush something in, that will affect existing themes in some way, without giving themers / theme maintainers a chance for feedback or even proper testing. Luckily (in this case) it wont make it into 1.22, so we have more time to discuss our best approach - also A11Y-wise.

**Comment by BWPanda** [comment]
I have some things to say about this, but need time to properly read all the latest comments and then respond. Since this didn't make feature freeze, seems I have a few months now to do so 😉

**Comment by BWPanda** [comment]
> @BWPanda What do you think of this?

I can see how trying to re-arrange all the fields is beyond the scope of this issue, so I like that you've simply added a heading for these checkboxes. However I think the wording for the two settings should be something like:

- [ ] Add a 'Title' field to comment forms
      *When disabled, comment titles will be automatically generated from the comment body.*
- [ ] Hide titles on comments
      *Comments will be displayed without a title.*

> I looked around and realized that quite a few contrib themes seem to override the comment.tpl.php file which seems to be a problem. I've been trying to write this so that it does not depend upon logic in the comment.tpl.php file.

@stpaultim Please see https://github.com/backdrop/backdrop-issues/issues/3822, specifically this part:

> ... almost every UI setting can already be overwritten by a theme (e.g. you can add custom titles to blocks, but all a theme has to do is hard-code a title in a template file and that title setting in the UI isn't going to work anymore). So unless we're going to change that behaviour across all of Backdrop, I'd say the answer here is that **themes can always override UI settings**. That's the way it works currently, that's the way Drupal 7 works, and that's what most people expect (based on the current state of things).

So trying to find a way to implement a UI setting that bypasses the theme's ability to override that setting, is the wrong way of looking at it.

I don't think we should be removing the title from the preprocess function, because then themes that still want it can't have it. Instead, we should add a boolean variable to the preprocess function that's available to template files where that value can be checked and the title displayed (or not) appropriately. E.g.:

```php
<?php if (!$hide_title): ?>
  <h3><?php print $title; ?></h3>
<?php endif; ?>
```

**Comment by indigoxela** [comment]
> we should add a boolean variable to the preprocess function that's available to template files where that value can be checked and the title displayed (or not) appropriately

That definitely looks like a better (and more BC) approach. :+1: 

Another thought: with regards to A11Y, I think it's not optimal to *completely remove* the title. Instead I'd like to suggest to "visually hide" the title, or turn it into a small icon. Putting the permalink to the date might be confusing (screen readers...) and TBH feels like abusing that comment header part.

Here's a mockup for a visually reduced but still present title:

![comment-title-as-icon](https://user-images.githubusercontent.com/842066/166235427-eed29682-e65b-47c5-83b7-b0ce9b908044.png)

With this approach, most changes could be done in CSS, based on a new class (by setting). So clashing with existing theme styles is less likely. :wink:

**Comment by docwilmot** [comment]
Re the hiding of the checkbox for custom title if title itself is hidden, I commented on the PR directly, but to reiterate here, I think *requiring* a custom comment title that however will not be shown is at best an extreme edge case. If you need a title for an admin list or whatever, you can auto generate from the comment body, like core does. I think if we hide the title checkbox, we should hide the custom title checkbox.

**Comment by stpaultim** [comment]
> What worried me here was that "last minute" is was attempted to rush something in...without giving themers / theme maintainers a chance for feedback or even proper testing

Sorry to give that impression (again). Yes, I started out exploring this issue to see if it might be simple and straightforward enough to get in. But, I was never committed to that idea and was just having fun with an issue that interested me. It was pretty clear to me that it would not get in, but I kept working anyway.

> Now things calm down again

I really don't understand this point. What wasn't calm about the work happening last night. I had momentum and was having fun working on an interesting issue.

**Comment by BWPanda** [comment]
I think @indigoxela was referring to comments such as:

> I had brief thoughts about getting this into 1.22, but I suspect that is not realistic - despite how close it is (or might be).

> If any core committers are still awake, this shouldnt be too far from getting in.

**Comment by stpaultim** [comment]
> So trying to find a way to implement a UI setting that bypasses the theme's ability to override that setting, is the wrong way of looking at it.

I don't think I was doing that at all. I'm don't think there was anything in my PR that tries to bypass the themes ability to override it nor would I have tried to do so. 

> Instead, we should add a boolean variable to the preprocess function that's available to template files where that value can be checked and the title displayed (or not) appropriately. E.g.:

I believe that this was the initial idea we experimented with. BUT, if I understand correctly, this solution will not work with existing contrib themes that have custom comments.tpl.php files, unless they update them. 

> Putting the permalink to the date might be confusing (screen readers...) and TBH feels like abusing that comment header part.

I'm not sure if this is true. I liked this idea, because it's a recognized UI pattern used in most major social media platforms. I'm quite sure that Facebook, Twitter, and LinkedIn all do this. 

On the otherhand, I think the little icon idea is also interesting and quite probably a better solution for contrib themes. 

I like this idea in this case, because as @indigoxela points out, it does not change anything in existing contrib themes. At worst, it replaces one thing (the title) with a permalink icon or word - if the site admin chooses this option. 

There is nothing to prevent contrib or custom themes from making other choices. We could make a different choice for Basis or Bartik if we want.

If I understand the idea correctly, this is how this idea works:

1) We conditionally add a class to the title field if the title is supposed to be hidden. 
2) We provide CSS that replaces the title with a little "link" icon that serves as the permalink.

**Comment by stpaultim** [comment]
I've updated the PR to try out (ROUGHLY) the suggestion by @indigoxela. It's a simpler version for now with some sloppy code, I need help. But, the purpose it to test the idea. 

This PR just does the following. 

1) If the "hide titles" option is selected, this PR adds a class to the title "hide_comment_title". 
2) It uses CSS to replace the title with the word "Permalink" (I would prefer to replace it with an icon, but I'm not sure how to replace it with an icon acts like a link using CSS). 
3) Improved titles of settings based upon input from @BWPanda 

PROBLEMS:

1) I had to put the CSS in the system.css file, because many themes have their own comments.css which overrides the core comments.css. This does not seem like the correct solution, but I need help coming up with a better idea. 
2) I'm using `text-indent: -99999px;` to hide the comment title. I'm open to better ideas. 
3) Following input from @docwilmot, I would like hide the "Custom Titles" field if the "Hide Titles" option is checked. But, if we do that, I think we need to make sure that custom titles are disabled. I haven't figured that out yet. 
4) Contrib themes that already provide special "Permalink" links, will now have two of them if titles are hidden. While not ideal, I think it's better than the alternatives. When we convert our "Permalink" link into an icon, that will be a little better. 

THINGS TO DO (or talk about):

1) We have the opportunity to do something different with Basis and Bartik, if we want. Right now, they are treated the same as every other theme. 
2) I think we should also provide some better default markup for this feature in the comments.tpl.php file in core. 

Looking for the following feedback:

1) How do to the things better that I mentioned above?
2) Do we like this general approach better than earlier ones?

We have plenty of time to talk about and work on this issue, no rush.

**Comment by indigoxela** [comment]
@stpaultim many thanks for working on this.

> It uses CSS to replace the title with the word "Permalink"

One concern: words added via CSS won't be translatable. I'd prefer to *not* add untranslatable strings.

> I had to put the CSS in the system.css file, because many themes have their own comments.css

Not sure if that's a good idea. If themes have their own comments.css, that's by intention. "Tricking" them to use the new code seems like the wrong approach. And system.css is also often unset in themes - as that file does too much. (And IMO system.css is the wrong file for this change.)

> I would like hide the "Custom Titles" field if the "Hide Titles" option is checked

I think, that's the wrong way round. I'd rather recommend to hide the "Hide Titles" checkbox, if "Custom Titles" is checked. As @docwilmot stated - it's unlikely that admins want to hide the comment subject, if they request to put content in them.

> I think we should also provide some better default markup for this feature in the comments.tpl.php

I'd recommend to not touch that file at all. And we don't need to do that, anyway.

> this ROUGHLY what you had in mind or did I misunderstand?

Very roughly it is. Just needs some tweaks here and there.

But maybe we should get some feedback first, if the CSS approach (icon) to prevent duplicate text finds consensus.

**Comment by olafgrabienski** [comment]
> I'm using text-indent: -99999px; to hide the comment title. I'm open to better ideas.

There is a CSS class `.element-invisible` in core to "hide elements visually, but keep them available for screen-readers". It's in `system.css`, btw:

https://github.com/backdrop/backdrop/blob/f389c984b234d1ebb2f32b6513d88464f1749fab/core/modules/system/css/system.css#L280

**Comment by stpaultim** [comment]
> Not sure if that's a good idea. If themes have their own comments.css, that's by intention. "Tricking" them to use the new code seems like the wrong approach. And system.css is also often unset in themes - as that file does too much. (And IMO system.css is the wrong file for this change.)

I said that this is probably a bad idea from the start. Telling me that it is a bad wrong approach is not helpful. I provided this solution with the a specific request for a better idea, because I'm stumped. But, I'm looking for a better solution that also accomplishes your goal of minimizing (or eliminating) the need for contrib maintainers to update their themes. 

I'm asking - where can we put this css for it to have any effect. If we decide to put it in comments.css, then this feature will not work with many contrib themes and we are not really any better off than we were before. 

Finding a way for this css to work is not tricking anyone to use the new code. No one made the decision to exclude this css from their theme, they made the decision to create their own comments.css file before this code existed and would likely have included this css in their files if it existed at that time. 

But aren't we looking for a solution that will work with existing contrib themes?

**Comment by BWPanda** [comment]
> I think, that's the wrong way round. I'd rather recommend to hide the "Hide Titles" checkbox, if "Custom Titles" is checked.

Perhaps this whole issue of the two checkboxes should be simplified and made into a single set of radio buttons:

**Comment titles**
- Custom
  *A title field is added to comment forms to allow for custom titles.*
- Auto-generated
  *No title field is shown, comment titles are automatically generated from the comment body.*
- Hidden
  *Comment titles are not displayed at all.*

**Comment by indigoxela** [comment]
> Perhaps this whole issue of the two checkboxes should be simplified and made into a single set of radio buttons

That's for sure worth a try. The current trouble is that the two checkboxes compete - radios would be mutually exclusive. Unless I miss something.

**Comment by quicksketch** [comment]
This seems like a new feature and we're past feature freeze for 1.22.0, so I'm bumping this to 1.23.0

**Comment by stpaultim** [comment]
Quick update:

In general, I think we may want to take some time to think on this issue - we have 4 months until the next feature freeze. As far as I can tell, none of our proposed plans of action will work without impacting contrib themes at all. I've opened a new topic to discuss in more detail the broader the issue of core changes that impact contrib themes = https://github.com/backdrop/backdrop-issues/issues/5614.

If anyone has specific ideas for implementing this feature in a way that doesn't have any impact on contrib themes, please share. I've expressed my concerns about the CSS method above, but maybe I missed something. 

> Perhaps this whole issue of the two checkboxes should be simplified and made into a single set of radio buttons:

I'd like to explore this idea now, while we consider the bigger issue. I may find time for a PR update that tests this idea, without solving the larger issues.

If anyone would like to submit their own PR for consideration, I'm fine with that.

**Comment by indigoxela** [comment]
> we may want to take some time to think on this issue - we have 4 months until the next feature freeze

Feature freeze means: no new items _after_ that date. It does not mean no new items _before_ that date. :smile:
It's seems a lot easier for all involved people (also core committers), if we don't wait until the last week before freeze - we can add new features to the new milestone any time.

@stpaultim I'd like to try the new UI (radios), but unfortunately no Tugboat demo has been created (and as expected, lots of comment related tests fail).
And I also left some comments on your PR.

**Comment by indigoxela** [comment]
Cool, Tugboat works again. This is what we have now. I like the UI!

![comment-setting-radios](https://user-images.githubusercontent.com/842066/167289186-e40f048d-953c-48d2-bc4a-d82b1c918bd6.png)

**Comment by indigoxela** [comment]
Currently the text rather suggests that hidden "titles" (comment subjects) are completely hidden.

If - and only if - we decide to show an icon instead, I'd suggest CSS like that - and suggest using svg, not png (explanation below):

```
.comment-subject-icon h3 .permalink {
  width: 1em;
  height: 1em;
  display: block;
  overflow: hidden;
  text-indent: -999px;
  background-color: currentColor;
  -webkit-mask-image: url("../../../misc/permalink.svg");
  mask-image: url("../../../misc/permalink.svg");
  -webkit-mask-size: cover;
  mask-size: cover;
}
```
If we use an svg as mask, and currentColor as the background, the icon will automatically show up in the current theme's link color.

If we decide to completely hide the subject, we should simply add the class element-invisible to the title (via preprocessor).

Another thought: we could *always* add a class to the title (subject). So themers can decide how to deal with the different types more easily.

**Comment by stpaultim** [comment]
@indigoxela - thanks for the excellent feedback and ideas. 

I've added a second PR, trying different things in each one for comparison. Neither one is probably the right mix just yet. 

At this moment, this is what each PR does:

4569-1

- Update hook to replaces old `comment_subject_field` boolean setting with `comment_title_options` setting (3 choices - auto, custom, hidden)
- If "Hide Titles" is selected, a css class `hide-comment-titles` is added to the comment titles, it uses CSS to hide comment titles on any theme that is displaying comment titles and replaces them with the word "permalink" that acts as a permalink. (If we go this route, we can swap out the word 'permalink' for the icon. 

4569-2

- Update hook to replaces old `comment_subject_field` boolean setting with `comment_title_options` setting (3 choices - auto, custom, hidden)
- If hidden is selected, it uses the comment.tpl.php file to hide the comment title in core, basis, and bartik. It does not have any effect on contrib themes using custom comment.tpl.php or comment.css files
- In BASIS, it uses CSS to add a permalink icon when the titles are hidden.
- In core and Bartik, it keeps the same old permalink link that was there by default.

I may experiment more with other options later. Neither PR is complete yet, but are experiments for comparison and works in progress. Ideas welcome. 

CURRENT QUESTIONS:

1) Labels - current using `comment title` in UI, variables and config. In past, the config setting was `subject field`. Is subject better than title?

**Comment by stpaultim** [comment]
I've updated the first PR to swap out the word "permalink" for the icon (see screenshot in previous comment). I think that this option is going to work pretty well for most contrib themes. It's not perfect, but possibly the best we will do without really jumping through hoops.

**Comment by indigoxela** [comment]
Ehem... `Call to undefined function dpm() ` ... @stpaultim I guess, your alternate PR needs a little cleanup. Let us know, when you think it's ready for testing.

**Comment by stpaultim** [comment]
dmp is removed. I'm not looking at tests and until we have decided on the final approach.

**Comment by stpaultim** [comment]
For anyone coming to this issue new. This issue currently has two PR's described in this comment: https://github.com/backdrop/backdrop-issues/issues/4569#issuecomment-1120588501

I think the next step is deciding which of these two approaches is the direction we prefer. I can then start to deal with details within whichever approach we decide on.

I will put this on agenda for two weeks from now.

**Comment by BWPanda** [comment]
I'm in favour of '4569-2 -- Template Solution'. This seems to me like the 'proper' way of doing things. Contrib themes that override core files should understand that they will need to manually update them should the core files change (assuming they *want* those changes, which isn't always the case). I also like how this solution doesn't assume everyone wants a permalink icon (which borders on a new feature/issue). Adding it for Basis is fine, as is leaving core/contrib to decide how they want to handle permalinks.

**Comment by argiepiano** [comment]
I see benefits and drawbacks with both approaches. The CSS approach will work without the need of modifying custom theme-provided templates, which is nice. If you are a site builder it would be confusing to click that option and finding that it doesn't "work". 

But this (CSS) approach is also unorthodox - typically theme builders expect to receive variables from the preprocessing layer and then make decisions based on those variables. Injecting a variable css class depending on a setting is very unusual (never seen this done by the preprocess layer).      

If the consensus is to go with the template solution, perhaps it'd be helpful to have a little warning in the content type configuration ui explaining that hiding the title may not work with all themes.

**Comment by BWPanda** [comment]
> perhaps it'd be helpful to have a little warning in the content type configuration ui explaining that hiding the title may not work with all themes.

No.
Most (all?) UI settings across Backdrop can be overridden by themes (e.g. add a custom title to your block, but if the current theme overrides the block template and removes or hard-codes the title, your custom title won't show). Do we add warnings to all settings? Because the fact that themes might override something is just the way Backdrop (and Drupal) work.

**Comment by indigoxela** [comment]
I'm fine with either approach. Both still need work.

Personally, I'd find a warning like "this might not work" rather confusing, so let's better not add it. Although I don't fully agree with @BWPanda - sure, that's how things work in D and B, but usually we try to be backwards compatible. :wink: If that's not fully possible, we try to make it as compatible as we can.

**Comment by olafgrabienski** [comment]
Changing the template seems to be the common approach in such a case. As a theme creator, if I add my own comment template, I know I'm overriding something, and I'm aware it can change, so I'm fine with the template approach (without warning).

**Comment by stpaultim** [comment]
Thanks for feedback. I understand that theme creators override things all the time. I have two concerns with making the changes in the theme template. 

1) Making contrib theme maintainers update their theme, when it might not be necessary. This is the lessor of my concerns and I can easily get by this one.
2) I'm uncomfortable with the user experience for site admins that check a box to "hide comment titles" and their comment titles are not hidden, because the contrib theme overrode the template. These site owners or admins may not understand that the feature is not working because of the specific contrib theme they are using. For them, the feature just doesn't work and it seems like a bug - or that maybe they are doing something wrong. 

A significant number of contrib themes override the comment.tpl.php file. If we are going to add a config option that simply won't work with many themes (as they exist today), I'm inclined to provide some kind of help text or warning to site owners/admins/users.

**Comment by stpaultim** [comment]
This issue was discussed at the end of the dev meeting today, without any clear decision regarding which direction to go. There was some "soft" sentiment in favor of the CSS solution and some understanding of the advantages of the template option. 

I would say that at this point, I'm softly leaning towards the CSS solution - BUT, understand the concerns and can easily be pushed in the other direction if I see a consensus moving in that direction.

**Comment by BWPanda** [comment]
> there are not any themes that chose to override this setting

I disagree. While themes may not be choosing to override *this* particular setting (because it didn't exist), by overriding a core template file they are effectively choosing to override all current *and future* settings/changes.

> I wonder if we don't have a responsibility to make sure that it has as minimal an impact as possible on existing contrib themes.

This is why we have release notes stating what changes have been made, and why we have a feature-freeze window between merging new features and releasing the new version - so users (and contrib maintainers) can test the new features and update their projects accordingly.

We mention in the release notes if changes have been made to core files outside the `/core` directory (e.g. `.htaccess`, `robots.txt`, etc.) and give people notice that they may want to copy those changes over to their versions of those files. If we do that with files that users *haven't* purposefully decided to override, why do themes that *have* decided to override core files get special treatment?

> I think we can provide this warning in a subtle and approrpriate way and it can be removed in the future, if we see that most contrib themes have made the necessary adjustment.

That's a rabbit hole I don't want to go down.

**Comment by stpaultim** [comment]
> That's a rabbit hole I don't want to go down.

Well, we can avoid this by using the CSS option, which works on all contrib themes. 

I can't think of any other recent examples where we added a feature that we knew from the outset that it would not work on many existing contrib themes. I just wonder if this is really a rabbit hole. It seems to me that it is pretty rare that are or would be adding options to core that we know from the outset will not work on many contrib themes.

**Comment by BWPanda** [comment]
I'm not sure of any features we've merged in this situation, but these two will have the same effect if/when merged:
- https://github.com/backdrop/backdrop-issues/issues/2351
- https://github.com/backdrop/backdrop-issues/issues/576

**Comment by docwilmot** [comment]
I prefer the second option as well, not the CSS option. We provide the ability to do this via code in the template, which is the right way to allow all themes to handle the option properly without forcing CSS and image files into the equation. We actually *use* this new ability in core themes, but since its optional, existing sites dont get affected unless the site builder wants it. If site builders with custom themes that are not yet coded to use this new ability ask why the UI option doesnt have an effect, we tell them the same thing we'd say if they ask why comment titles dont appear on some themes: your theme doesnt support it (yet).

**Comment by olafgrabienski** [comment]
I'd like to understand why so many contrib themes override Backdrop core's `comment.tpl.php`. To mention *one possible* reason, theme maintainers may have overridden that template file just because they missed the setting discussed in this issue (and/or similar ones). I guess, in this case many would be happy to change the contrib template file to make the setting work in their theme.

**Comment by indigoxela** [comment]
> I'd like to understand why so many contrib themes override Backdrop core's comment.tpl.php

I do it because that template isn't very good for styling. I'm one of those who does override that template file when I need comments. And I already stated, that I'm fine with either - CSS or template solution. And I'm aware that I very likely have to update in my contrib themes, no matter what the core decision will be. Possibly, because I'm pretty opinionated when it comes to styles. :wink: But I'm not sure if that's relevant here.

IMO it's OK if the setting doesn't work out of the box in themes that override the template and/or comment styles.
But the core change shouldn't *break* comment styles in existing themes (contrib or custom) - that's one thing we have to take care of (and we will).

**Comment by olafgrabienski** [comment]
> I usually do it because I want to remove comment subjects :)

Interesting! So, if we pick the CSS solution, the theme override would be needless and could probably be changed or removed, or its character should at least be documented. It that's the case, and if theme maintainers are willing to do such changes (I guess, they are) I see no advantage of the CSS solution in comparison to the template solution. Quite the contrary, if I don't miss anything, the template solution would probably motivate theme maintainers to remove unnecessary `comment.tpl.php` overrides.

**Comment by stpaultim** [comment]
Here is a list of all the themes on BackdropCMS.org

Themes with "x" have overwritten the comment.tpl.php ile

Name                   Installs         Maintainer
=====                   =====         =========
Opera                           35        Tim Erickson
Scenery                        19   x    Indigoxela
Axioma                         50         Pavel
Bootstrap 5 Lite            62    x   DocWilmot
Seventy                         12        Olaf
Monochrome                 18   x   Indigoxela
Snazzy                          62   x    Indigoxela
Pelerine                        70    x    Indigoxela
Lateral                         128   x     Indigoxela
Shasetsu                       89         Olaf
Thesis                           28         Olaf
Lumi                              98         BWPanda
Tatsu                             76          Tim Erickson
Corporate Kiss            152          Vincent Stemen
minicss                          26         Vladimir
Borg                               61   x    Jen Lampton
Basis (Contrib)                 9         NONE
Pretty in Purple              10    x   Jen Lampton
Bootstrap Lite              194    x    Tim & Robert
ZURB Foundation 6       32   x     Biolithic
GDI Tribute                      5    x     Biolithic
Teamwork 15                   1    x     Tim Erickson 
Wormhole                        4    x     Ezenwa Isiogu
News Arrow                   21    x     Biolithic
Mero                              50    x     DocWilmot
Colihaut                         62    x     DocWilmot
Summer Fun                 16    x     Biolithic
Afterlight Tribute            28    x     Afterlight Tribute
Materialize                     58    x     Biolithic
Zurb Foundation 5           1    x     Biolithic
Cleanish Theme             58   x     DocWilmot

Just in caser were relying on theme maintainers to impelment this change in their themes. 

NOTING: Even if theme developers implement the change, how many users regularly apply non-security related updates to themes? I'm not sure I would. If I'm happy with my theme, why would I risk updating it without an urgent reason?

How many people that use Backdrop expect that they need to update their theme to get a new core feature?

**Comment by indigoxela** [comment]
> How many people that use Backdrop expect that they need to update their theme to get a new core feature?

Not sure about that, but I'd assume that if something unexpected happens, the first thing to check is if everything's up to date.
And also: if people update core, they probably also consider contrib, also themes. So IMO that's no major concern.

People who never update unless it's an urgent security update, probably also wouldn't be interested in new core features - only guessing, though. :wink:

**Comment by VasasA** [comment]
If we can check the compatibility of the current theme with the "Hide title" option: If it is not compatible, we can disable the option, and display a text below the option label: "Current theme needs to be updated for this option."

**Comment by indigoxela** [comment]
> If we can check the compatibility of the current theme with the "Hide title" option...

I don't think that's feasible, though. How would we define "compatibility"? Nothing breaks. Currently the consensus seems to go the CSS-only route - all/most themes will be compatible.

**Comment by stpaultim** [comment]
@quicksketch expressed a pretty clear preference to updating the template in the meeting today. I think we need to focus on that solution. 

As someone that is pretty concerned about how this might impact users, I'd like to pursue either one of these options:

1) Provide some information in the UI regarding why this feature may not be working with some themes. I open to keeping this pretty subtle. 

2) Or providing PR's for all the unsupported themes that will address this problem and then deciding if the bug squad has authority to merge this fix. It looks to me lilke there are currently 8 themes without an active maintainer that would be effected by this and I think creating a PR to fix it for these themes is pretty easy. It will then be a policy decision regarding whether or not this is a good idea.

**Comment by indigoxela** [comment]
Yay, a decision! Be it a change in the template file then. :wink: Let's go on with the work now.

First suggestion for the implementation: don't use the constants directly in the tpl.php files, but set a proper variable (or variables) in the preprocessor.

**Comment by stpaultim** [comment]
Heads up to these maintainers:

- @docwilmot - You have 4 themes that will be affected by this change.
- @indigoxela - You have 6 themes that will be affected by this change.
- @jenlampton - You have 2 themes that will be affected by this change.
- @stpaultim - I have 1 theme that will be affected by this change.
- @clickbox - You have 1 theme that will be affected by this change.
- @biolithic - You have 7 themes that will be affected by this change.

I have created a sample PR for the Materialize theme that does the minimal fix required to support this new feature. 
https://github.com/backdrop-contrib/materialize/issues/8

Once we are closer to committing this new feature, I will create PR's against some of the other themes that are not actively maintained and discuss options for getting these PR's committed. Keeping in mind that most site owners will have no way to understand that this feature is not working because of the theme. To most site editors, it will appear as if core is broken and I'd like to do my best to avoid this situation.

**Comment by stpaultim** [comment]
The PR is ready for review and testing again. 

I've added this warning which I expect to be controversial based upon previous comments. 

![image](https://user-images.githubusercontent.com/3144571/175794594-e9defd6e-a5b1-4e8c-ac1c-d42acc228d5f.png)

Here is my argument:

1) I understand that not every contrib theme supports every feature in core and we can't necessarily expect that. 
2) However, I believe that most cases a theme not supporting a feature means that the feature looks bad and it's fairly easy to trace the problem to the theme and understand the cause - with an obvious solution, find another theme. 
3) I think that this case is different because if the theme does not support the feature, the feature just doesn't work. The user clicks on a checkbox in core and nothing happens. There is not obvious clue to the user that this is a problem with the theme. It will appear to the user that a core feature is just not working and there is no easy way for them to understand that switching themes might fix the problem. 

While I am sure I can find LOTS of features that look bad when used with some contrib themes. I am not sure I can think of any examples, where the feature just doesn't work in contrib themes without any clue as to why it's not working. 

We had a possible solution to prevent this problem, but we decided that this solution was a bit messy and incovenient for core/contrib developers. I accept this decision. 

I'd like to offer the same considertion to the end users for whom this choice may lead to very confusing and unexpected behaviors in core. 

I welcome other examples of core features that just don't work at all in at least several contributed themes without any obvious clues to the cause of the problem. While, I recognize that not alll themes will support all core features. I think it's reasonable to provide at least a small clue to end users about a new feature that could be very confusing when it just doesn't work. For me, the end goal here is to make life less frustrating for any users that might try this feature and not understand why it's not working.

This warning is only a few words long and easy to implement. It will potentially help some users avoid unnecessary confusion and can be removed quite easily at a later date if it is no longer necessary (we could put a @TODO note to review the situation later, but we don't have to). If it doesn't get removed, that isn't a problem for anyone either. 

I know that there is concern that this will lead to lots of other features with similar warning labels, which I do not believe is true. I believe that this is a very rare situation where we have a new feature that is known to be incompatible with so many contributed themes in such a mysterious way.

**Comment by indigoxela** [comment]
> The PR is ready for review and testing again.

I already plagued you with some more suggestions. :wink: Overall it still works for me, BTW. I'm just trying to help polishing it.

> I think that this case is different because if the theme does not support the feature, the feature just doesn't work. ... It will appear to the user that a core feature is just not working

That's a valid point. I'm fine with that tiny sentence, leaving text suggestions to native speakers.

> ... a new feature that is known to be incompatible with so many contributed themes in such a mysterious way.

Currently lots of theme's don't support the new feature yet, but the situation will be different, when the next B minor comes out in September. :wink: Promised.

**Comment by stpaultim** [comment]
I've address some of @indigoxela comments and need feedback on others (see PR). 

Specific things to look at or comment on. 

1) core/modules/comment/tests/comment.test (line 199) - No error here, but I made changes here that I'm a little uncertain about (a lot uncertain about). 
2) Use of constants. 
3) New classes.

**Comment by indigoxela** [comment]
@stpaultim many thanks for your updates and comments and questions.

One thing: currently the comment module provides the svg, but doesn't use it at all. Only basis provides css to use it. Shouldn't the image go to Basis instead?

**Comment by stpaultim** [comment]
> One thing: currently the comment module provides the svg, but doesn't use it at all. Only Basis provides css to use it. Shouldn't the image go to Basis instead?

This is something I wanted feedback on and I think I asked about it somewhere in the past. I don't know what the precidence is in a case like this. Here is the situation:

1) The core module and Bartik theme insert the word permalink that works as a permalink, regardless of whether titles are shown or not. I don't think we should remove the existing permalink or add an icon where an existing link already exists. 

2) When hiding titles in Basis, there is no remaining permalink so we replace it with an icon instead of the word "permalink." 

I could have put the new icon directly in Basis, where is it needed right now. However, I think that in the future having the link available for all themes would be more useful so I included it in the core comment module. I don't feel strongly about this, but unless others object, I like the idea of adding the link to the core module for use by any theme, even if it's not used by the core comment.tpl.php file.

Thoughts?

**Comment by stpaultim** [comment]
> May I ask that we keep changes here to a minimum? This issue is about allowing comment titles to be hidden, so let's focus on that here. If we then want Basis to have a permalink, especially when it's changing from a string to an icon, that's a design decision that should be discussed in a separate issue.

Currently titles serve as Permalinks. Back in May, I did a PR that removed the title (and permalink) without replacing it and got this feedback from @docwilmot

> I'd say thats an essential feature really, so I'm bumping to needs work.
https://github.com/backdrop/backdrop-issues/issues/4569#issuecomment-1114293958

The new (replacement) permalink has been part of every version since then. It seems to me that removing it at this stage would be both removing something that currently exists and was previously requested as an "essential" feature. 

It is only coming up again now in terms of how to best implement it. If the discussion about where to put the icon is creating unnecessary complexity, I'll put it in Basis where I think it is needed and we can leave any other discussions for the future.

**Comment by docwilmot** [comment]
> If we then want Basis to have a permalink, especially when it's changing from a string to an icon, that's a design decision that should be discussed in a separate issue.

I felt that not having a permalink (which is what removing the title would accomplish) would be introducing a bug here, hence my initial reaction. 

I like the UI of this current PR. In Basis the icon is a good compromise and Bartik keeps its "permalink" string. LGTM.

**Comment by BWPanda** [comment]
I made a few suggestions too.

**Comment by stpaultim** [comment]
Thanks everyone for the patient and thoughtful feedback. Apologies for the sloppy mistakes. Once again, I think I've incorporated all the current feedback.

**Comment by indigoxela** [comment]
Oh, almost forgot to mention: I'd prefer to have (updated) functional tests for the new functionality (as there's a new form item). Any volunteers?

**Comment by jenlampton** [comment]
I notice that we're adding a new link icon, but it's a SVG. Everywhere else we use SVGs in core we also have a png fallback, as follows:
```
div.messages.info:before {
  background-image: url(../../../misc/message-32-info.png);
  background-color: #0074bd;
}
/* SVG overrides */
div.messages.info:before {
  background-image: linear-gradient(transparent, transparent), url(../../../misc/message-info.svg);
}
```
We should do the same here.

**Comment by stpaultim** [comment]
@jenlampton has agreed to help with tests. I asked if we could do it together, so I can learn more and she agreed to work on this collaboratively just after Office Hours next week. 

If anyone else would like to learn more about writing tests with a specific practical example, this would be a great time to drop in.

**Comment by stpaultim** [comment]
@jenlampton helped me write some new tests. I've pushed those up. 

Still to do, add upgrade path for contrib module.

**Comment by stpaultim** [comment]
I was getting random test failures, but current PR is all green. 

I would really like someone good with tests to review lines

core/modules/comment/tests/comment.test 447-464

These tests are not failing, but are they actually testing anything useful?

**Comment by indigoxela** [comment]
@stpaultim it seems like there's at least one change request on you PR (from July 2nd), which hasn't been addressed yet. So setting back to "needs work".

**Comment by indigoxela** [comment]
Re the automated tests: I think they're doing what they should do. :+1:

**Comment by stpaultim** [comment]
@indigoxela - Thanks for the feedback on automated tests. Although, I'm getting a failure now that seems kinda random. 

I think I've implemented all the feedback I've gotten, including some recent thoughts by @klonos. Changes are significant enough that a think another round of (light?) testing would be helpful. 

Also, in need of code review.

**Comment by indigoxela** [comment]
One thing came to my mind: what about contrib module that use "comment_subject_field" in custom content types? There seem to be quite a lot:

https://github.com/search?q=org%3Abackdrop-contrib+comment_subject_field&type=code

We can't (to my understanding) catch that with the update hook, as these modules could get installed *after* the update hook has run.

**Comment by stpaultim** [comment]
> what about contrib modules that use "comment_subject_field" in custom content types, demo content or functional tests?

Hmmmm - This seems like a problem. I don't have time to think about this tonight, but it seems like something we need to address.

**Comment by klonos** [comment]
Good catch @indigoxela 👍🏼 

Seems that if we should either not change the setting name, or add some "mapping" for backwards compatibility.

**Comment by klonos** [comment]
> I think I've implemented all the feedback I've gotten, including some recent thoughts by @klonos. ...another round of (light?) testing would be helpful.

For those not following the comments in the PR thread, this includes the following change:

Before:
![image](https://user-images.githubusercontent.com/2423362/185806999-e3f2bc07-463d-48d3-86f5-39fb2fe377c6.png)

After:
![image](https://user-images.githubusercontent.com/2423362/185807063-aae58e55-8721-4a29-b07b-f4a9429bd3f4.png)

**Comment by indigoxela** [comment]
So, what has to happen to make this BC?

- Back to `comment_subject_field` (don't switch the setting name)
- For the (now) three options provide values that also work with the previous true/false values
- Adapt the update hook accordingly

Now brainstorming for the constants/settings:

Old (current stable):

- `comment_subject_field = false` means the subject/title is generated
- `comment_subject_field = true` people get an input field to set a custom text

New:

- `comment_subject_field = COMMENT_TITLE_AUTO`, COMMENT_TITLE_AUTO = 0 (evaluates to false)
- `comment_subject_field = COMMENT_TITLE_CUSTOM`, COMMENT_TITLE_CUSTOM = 1 (evaluates to true)
- `comment_subject_field = COMMENT_TITLE_HIDDEN`, COMMENT_TITLE_HIDDEN = 2 (evaluates to true)

So the third one could be problematic. Instead of hiding, the form item is provided, which differs to previous behavior. Any ideas?

We could do stricter checks like `'comment_subject_field' === ...` to catch that. Other ideas? Am I missing something?

**Comment by indigoxela** [comment]
@klonos before this comes: yes, the setting name doesn't actually reflect what it does now, and the constants are back to numbers. IMO backwards compatibility has precedence here. Breaking stuff just for nice names IMO isn't the way to go.

**Comment by indigoxela** [comment]
Another idea: keep the current constant strings, provide a helper function to evaluate, whether the input field should be shown or the comment subject should be hidden.

In that helper function check all possible values true/false/"auto"/"custom"/"hidden".

**Comment by yorkshire-pudding** [comment]
I'm not sure I can offer any additional help on the solution, but happy to test when needed

**Comment by stpaultim** [comment]
I've added this problem with the variable name to the agenda for the next dev meeting, in case we don't come up with a great solution by then. Flagging @backdrop/core-committers for input. See: https://github.com/backdrop/backdrop-issues/issues/4569#issuecomment-1221928520

**Comment by klonos** [comment]
> Other ideas? Am I missing something?

I was thinking that something like #5741 perhaps, that would allow us to safely rename the current setting (while "redirecting" contrib that still uses the old name to work the same + logging a deprecation message to help contrib spot these issues and fix them in future versions). I'll need to think about it some more, but for now I have raised that other issue, so to not block things here.

**Comment by indigoxela** [comment]
> I was thinking that something like https://github.com/backdrop/backdrop-issues/issues/5741 perhaps,

Sure, actually a good idea, but... in that case this one here probably won't make it into 1.23 - feature freeze is nearing...

**Comment by klonos** [comment]
I would not be happy but OK with the weird/non-intuitive config setting names in this PR, if we made sure that we added a `@todo` to return and fix things once #5741 is implemented. That way this feature here won't be blocked.

Having said that, I'd like to know what others - especially core committers think about all this. Perhaps they have totally different ideas/solutions that we have not considered yet.

**Comment by jenlampton** [comment]
I'm throwing captcha (uses the old value in a test) and comment_hide_subject (uses the value in the module itself) and config recipes (which installs new content types) to see what kind of damage this change does when left as-is. I've applied the patch and run the update before installing these modules. 

This is the results when tests reference the wrong vaue:

<img width="1390" alt="Screen Shot 2022-08-25 at 3 05 12 PM" src="https://user-images.githubusercontent.com/397895/186777132-5aae59ec-6adc-46a5-b838-9072b78a279d.png">

This is what I get from the comment_hide_subject module (which may not be a real problem. why would you add this module after the feature was already in core?)

<img width="1185" alt="Screen Shot 2022-08-25 at 2 48 07 PM" src="https://user-images.githubusercontent.com/397895/186775100-12d1414f-3118-45c2-9813-f3b08b47ff0b.png">

There are no ill effects of config recipes that reference the wrong value in config. Everything continues to work as expected.

**Comment by bugfolder** [comment]
> which may not be a real problem. why would you add this module after the feature was already in core?

Not a problem. This PR disables the module if it's already installed, and as soon as 1.23 is released I'll do a bugfix release to the module that notes [in its README](https://backdropcms.org/project/comment_hide_subject) that it's no longer needed.

**Comment by stpaultim** [comment]
We discussed the issue raised by @indigoxela extensively in the dev meeting today. @quicksketch seemed to be of the opinion, that the best thing we could do was the create the helper function suggested by @indigoxela. But, he was unconvinced that this was really necessary. The end decision was to: 

1) Run some experiments to get a sense of how severe the problem might be
2) Unless the tests go poorly, to not worry about the backwards compatibility of this specific change as it doesn't really do any harm.

@jenlampton reported on some experiments she ran (see previous comment - two comments ago). As I understand it, the mere failure of these contrib tests is not enough to block this PR as it doesn't really prevent the modules from working. The tests can be fixed anytime without breaking any sites or functionality. 

If my understanding is correct, this module is ready for testing and code review again.

**Comment by indigoxela** [comment]
> As I understand it, the mere failure of these contrib tests is not enough to block this PR as it doesn't really prevent the modules from working

That's an interesting interpretation of "backwards compatibility", though. :stuck_out_tongue_winking_eye: Fixing tests takes time, too. Do we consider everything "BC", that doesn't completely break existing sites?

**Comment by indigoxela** [comment]
@stpaultim I've left another comment on the PR (the constants are probably obsolete, now that we use human readable values).

Did anyone already check, it this setting name change might be problematic with config syncs?

Another thought: contribs (like for example recipes) have to support both setting names, right? The next minor is an optional update, so contribs have to support both values - is that correct? Is there a best practice for supporting both?

**Comment by stpaultim** [comment]
> That's an interesting interpretation of "backwards compatibility", though. stuck_out_tongue_winking_eye Fixing tests takes time, too. Do we consider everything "BC", that doesn't completely break existing sites?

I don't know if you had time to watch this portion of the meeting, but I encourage it if you are interested in this topic. Others could do (and did) a better job of explaining WHY they considered this an acceptable change. To be honest, I did not follow all of the discussion. 

I can point out that during the discussion we referred to the "API Compatibility" section of this page to help clarify what kinds of changes are acceptable. https://backdropcms.org/releases

**Comment by herbdool** [comment]
I just watched the meeting. The API compatibility policy doesn't guarantee backwards compatibility. There was discussion on how many contrib modules are impacted, and whether a helper function is required. The idea of the helper function would be to convert the old value to the new when someone resaves a content type.

There was consensus not to use the old config name but give it a new value since that would confuse things. So the new name stays (and I agree). I think I understood that using constants was fine but instead of the constants holding numbers 0,1,2 they could hold strings that are more understandable of what they do.

**Comment by indigoxela** [comment]
Let's not get distracted here by policy discussions... :wink:

Again:

- Did anyone already check, if this setting name change might be problematic with config syncs?
- Do contribs (like for example recipes) have to support both setting names?

**Comment by herbdool** [comment]
1. I haven't tested this particular config sync but it isn't the first time that a core config changes. Site admins need to make sure they export the active config to staging (or however they usually sync) after the upgrade so that they don't wipe out live config changes.
2. A helper function could help with recipes so that after they are imported, the old setting is converted to the new on content type saving. Or recipes may need to focus just on the latest release (with a disclaimer perhaps). Because, as per the BC policy, there is no guarantee with configuration.

I recall new settings like automatic comment closer which went over smoothly. Not exactly the same case here but I don't think it'll be disruptive.

I had taken a look at which contrib modules mention the old setting. A couple of them I'm the maintainer so I'll be updating the tests on those.

**Comment by stpaultim** [comment]
I'm doing some testing with recipes. Recipes simple create this setting in config files when the content type is generated. Based upon my testing, after this change the correct setting would be added to the content type any time that the content type is changed. This would leave an unused setting in the recipe config file, but I don't think it does any damage and there is no need for the recipe to support it. 

I've already removed this setting from one of my recipes, since this setting is not required and will be created again once anyone saves the content type (either before or after the possible change to core). The only time it would be important to have this setting correct in the recipe would be if the recipe expected the setting to be something other than the default set in core. 

None of my recipes expect that, so I don't think it's a problem. I'll work to remove this setting from all my recipe modules, to keep things clean.

**Comment by stpaultim** [comment]
Update. After making changes to modules I maintain, this list is now down to 9 items. 
https://github.com/search?q=org%3Abackdrop-contrib+comment_subject_field&type=code

4 are tests
2 are json files or OG sample content (code I created, but do not maintain). I will submit a PR to fix these.
1 is a reference to the module that this PR will deprecate or replace
2 are json files for content types created by other modules

I'll submit PR's to these two modules should this PR get merged. But, based upon my testing, there is really no harm in leaving the settings in the json file as they are.

**Comment by herbdool** [comment]
I think it looks good now. I also tested all options. And others tested with other themes and contrib. I don't think there are any blockers. Any changes to contrib can be made after.

**Comment by quicksketch** [comment]
There's a little work needed on the update hook: https://github.com/backdrop/backdrop/pull/4068#pullrequestreview-1089354599

I'll work to clean this up a little.

**Comment by quicksketch** [comment]
@stpaultim I filed a PR against your PR at https://github.com/stpaultim/backdrop/pull/4.

**Comment by stpaultim** [comment]
@quicksketch - I merged your PR as it was.

**Comment by quicksketch** [comment]
Thanks @stpaultim. I found one more thing in the update hook: https://github.com/backdrop/backdrop/pull/4068#pullrequestreview-1089391132

**Comment by quicksketch** [comment]
I think this is now ready to go.

**Comment by quicksketch** [comment]
I've merged https://github.com/backdrop/backdrop/pull/4068 into 1.x for 1.23.0. Thanks everyone for the extensive background research, options exploration, and iterations on implementation. I remember comments having a title was a pet-peeve of Lullabot founder Jeff Robbins nearly 15 years ago, and now you can finally turn them off!

https://github.com/backdrop/backdrop/commit/85025fc263261a0eb35f4d9eeb8e29d971f8a813 by @stpaultim, @VasasA, @bugfolder, @BWPanda, @argiepiano, @docwilmot, @indigoxela, @olafgrabienski, @klonos, @jenlampton, and @quicksketch.

Thanks everyone!
